File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/VanishInComponent.cpp
#include <Altseed.h>
#include <gtest/gtest.h>
#include <memory>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class ObjectSystem_VanishInComponent : public EngineTest
{
	class VanishingComponent 
		: public asd::Object2DComponent
	{
		int time = 0;

		asd::Object2D::Ptr vanishedObject = nullptr;
		int etime = 0;

	public :
		VanishingComponent(asd::Object2D::Ptr vanishedObject, int etime)
		{
			this->vanishedObject = vanishedObject;
			this->etime = etime;
		}

	protected:
		void OnUpdate()
		{
			if (time == etime)
			{
				vanishedObject->Vanish();
			}

			time++;
		}
	};

	std::shared_ptr<asd::Layer2D> layer = nullptr;
	std::shared_ptr<asd::Object2D> obj = nullptr;

public:
	ObjectSystem_VanishInComponent(bool isOpenGLMode)
		: EngineTest(asd::ToAString("ObjectSystem_VanishInComponent"), isOpenGLMode, 40)
	{
	}

protected :
	void OnStart()
	{
		auto scene = std::make_shared<asd::Scene>();
		layer = std::make_shared<asd::Layer2D>();
		obj = std::make_shared<asd::TextureObject2D>();

		asd::Engine::ChangeScene(scene);
		scene->AddLayer(layer);
		layer->AddObject(obj);
	}

	void OnUpdating()
	{
		if (GetTime() % 2 == 0)
		{
			auto temp = obj;
			obj = std::make_shared<asd::TextureObject2D>();
			layer->AddObject(obj);
			obj->AddComponent(std::make_shared<VanishingComponent>(temp, 3), asd::ToAString("v").c_str());
		}
	}
};


ENGINE_TEST(ObjectSystem, VanishInComponent)

File: Dev/unitTest_Engine_cs/ObjectSystem2D/VanishInComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace unitTest_Engine_cs.ObjectSystem2D
{
	class VanishInComponent : EngineTest
	{
		class VanishingComponent : asd.Object2DComponent
		{
			int time = 0;

			asd.Object2D vanishedObject = null;
			int etime = 0;

			public VanishingComponent(asd.Object2D vanishedObject, int etime)
			{
				this.vanishedObject = vanishedObject;
				this.etime = etime;
			}

			protected override void OnUpdate()
			{
				if (time == etime)
				{
					vanishedObject.Dispose();
				}

				time++;
			}
		}

		asd.Layer2D layer = null;
		asd.Object2D obj = null;

		public VanishInComponent()
			: base(40)
		{
		}

		protected override void OnStart()
		{
			var scene = new asd.Scene();
			layer = new asd.Layer2D();
			obj = new asd.TextureObject2D();

			asd.Engine.ChangeScene(scene);
			scene.AddLayer(layer);
			layer.AddObject(obj);
		}

		protected override void OnUpdating()
		{
			if(Time % 2 == 0)
			{
				var temp = obj;
				obj = new asd.TextureObject2D();
				layer.AddObject(obj);
				obj.AddComponent(new VanishingComponent(temp, 3), "v");
			}
		}
	}
}
